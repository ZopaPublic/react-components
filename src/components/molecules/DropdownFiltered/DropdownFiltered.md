![BETA](https://img.shields.io/badge/BETA-0.0.4-yellow.svg)

### Summary

`<DropdownFiltered />` is a custom dropdown with a filter applied on typing.

It wrap four components:

- [`<SizedContainer />`](/#/Components/Layout/SizedContainer)
- [`<InputText />`](/#/Components/Atoms/InputText)
- [`<Label />`](/#/Components/Atoms/Label)
- [`<ErrorMessage />`](/#/Components/Atoms/ErrorMessage)

And uses [paypal/downshift](https://github.com/paypal/downshift) internally.

How it works:

- The filtered text is always display in an input text.
- The selected item is passed to the onChange function.

Also note that:

- You can set a default value ( see the examples üëáüèª )
- Options lists closes when clicking outside of it
- All props from [paypal/downshift](https://github.com/paypal/downshift) can be applied to it

### Todos

- Improve code readability
- Simplify prop API ü§Ø
- Test filter functionality

### Examples

- Standard

```jsx
import { DropdownFiltered } from '@zopauk/react-components';

const items = [{ value: 'apple' }, { value: 'pear' }, { value: 'orange' }, { value: 'grape' }, { value: 'banana' }];

<DropdownFiltered
  label="Fruits"
  items={items}
  name="fruit"
  onChange={selection => selection && console.log(`You selected ${selection.value}`)}
  inputProps={{
    placeholder: 'Select a fruit...',
    name: 'fruits1',
  }}
/>;
```

- With default value

```jsx
import { DropdownFiltered } from '@zopauk/react-components';

const items = [{ value: 'apple' }, { value: 'pear' }, { value: 'orange' }, { value: 'grape' }, { value: 'banana' }];

<DropdownFiltered
  label="Fruits"
  items={items}
  name="fruit-default"
  defaultSelectedItem={{ value: 'orange' }}
  onChange={selection => selection && console.log(`You selected ${selection.value}`)}
  inputProps={{
    placeholder: 'Select a fruit...',
    name: 'fruits2',
  }}
/>;
```

- Long list with a `max-height` applied to it

```jsx
import { DropdownFiltered } from '@zopauk/react-components';

const items = [
  { alpha2: 'GB', value: 'British' },
  { alpha2: 'AO', value: 'Angolan' },
  { alpha2: 'AI', value: 'Anguillan' },
  { alpha2: 'AG', value: 'Antiguan/Barbudan' },
  { alpha2: 'AR', value: 'Argentine' },
  { alpha2: 'AM', value: 'Armenian' },
  { alpha2: 'AW', value: 'Aruban/Dutch' },
  { alpha2: 'TD', value: 'Chadian' },
  { alpha2: 'GG', value: 'Channel Islander Guernsey' },
  { alpha2: 'CL', value: 'Chilean' },
  { alpha2: 'CD', value: 'Congolese Dem Rep Of Congo' },
  { alpha2: 'CG', value: 'Congolese Rep Of Congo' },
  { alpha2: 'CK', value: 'Cook Islander' },
  { alpha2: 'CZ', value: 'Czech' },
  { alpha2: 'DK', value: 'Danish' },
  { alpha2: 'DJ', value: 'Djiboutian' },
  { alpha2: 'DM', value: 'Dominican Dominica' },
  { alpha2: 'KN', value: 'Kittitian/Nevisian' },
  { alpha2: 'KP', value: 'Korean Dem People Rep' },
  { alpha2: 'KR', value: 'Korean Republic' },
  { alpha2: 'KW', value: 'Kuwaiti' },
  { alpha2: 'KG', value: 'Kyrgyzstani' },
  { alpha2: 'IT', value: 'Italian' },
  { alpha2: 'LA', value: 'Lao/Laotian' },
  { alpha2: 'MH', value: 'Marshallese' },
  { alpha2: 'MQ', value: 'Martinican' },
  { alpha2: 'MR', value: 'Mauritanian' },
  { alpha2: 'MU', value: 'Mauritian' },
  { alpha2: 'MX', value: 'Mexican' },
  { alpha2: 'ES', value: 'Spanish' },
  { alpha2: 'SZ', value: 'Swazi' },
  { alpha2: 'SE', value: 'Swedish' },
  { alpha2: 'CH', value: 'Swiss' },
  { alpha2: 'TJ', value: 'Tajikistani' },
  { alpha2: 'TZ', value: 'Tanzanian' },
  { alpha2: 'TH', value: 'Thai' },
  { alpha2: 'TL', value: 'Timorese' },
  { alpha2: 'TG', value: 'Togolese' },
  { alpha2: 'TK', value: 'Tokelauan' },
  { alpha2: 'TO', value: 'Tongan' },
  { alpha2: 'TN', value: 'Tunisian' },
  { alpha2: 'TR', value: 'Turkish' },
  { alpha2: 'TM', value: 'Turkmen' },
  { alpha2: 'TV', value: 'Tuvaluan' },
  { alpha2: 'UG', value: 'Ugandan' },
  { alpha2: 'UA', value: 'Ukrainian' },
  { alpha2: 'UY', value: 'Uruguayan' },
  { alpha2: 'UZ', value: 'Uzbekistani' },
  { alpha2: 'VE', value: 'Venezuelan' },
  { alpha2: 'VN', value: 'Vietnamese' },
  { alpha2: 'ZW', value: 'Zimbabwean' },
];

<DropdownFiltered
  label="Nationality"
  items={items}
  name="nationality-default"
  onChange={selection => selection && console.log(`You selected ${selection.alpha2}`)}
  optionsListMaxHeight="250px"
  inputProps={{
    placeholder: 'Select a nationality...',
    name: 'nationality1',
  }}
/>;
```

- With an error message

```jsx
import { DropdownFiltered } from '@zopauk/react-components';

const items = [{ value: 'apple' }, { value: 'pear' }, { value: 'orange' }, { value: 'grape' }, { value: 'banana' }];

<DropdownFiltered
  label="Fruits"
  items={items}
  name="fruit-error"
  onChange={selection => selection && console.log(`You selected ${selection.value}`)}
  errorMessage="Ops! There is an error!"
  inputProps={{
    placeholder: 'Select a fruit...',
    name: 'fruits3',
  }}
/>;
```

- Valid state

```jsx
import { DropdownFiltered } from '@zopauk/react-components';

const items = [{ value: 'apple' }, { value: 'pear' }, { value: 'orange' }, { value: 'grape' }, { value: 'banana' }];

<DropdownFiltered
  label="Fruits"
  items={items}
  name="fruit-isValid"
  onChange={selection => selection && console.log(`You selected ${selection.value}`)}
  inputProps={{
    placeholder: 'Select a fruit...',
    name: 'fruits4',
  }}
  isValid={true}
/>;
```

- Disabled

```jsx
import { DropdownFiltered } from '@zopauk/react-components';

const items = [{ value: 'apple' }, { value: 'pear' }, { value: 'orange' }, { value: 'grape' }, { value: 'banana' }];

<DropdownFiltered
  label="Fruits"
  items={items}
  name="fruit-disabled"
  onChange={selection => selection && console.log(`You selected ${selection.value}`)}
  inputProps={{
    disabled: true,
    placeholder: 'Select a fruit...',
    name: 'fruits5',
  }}
  defaultSelectedItem={{ value: 'orange' }}
/>;
```

- With a short size

```jsx
import { DropdownFiltered } from '@zopauk/react-components';

const items = [{ value: 'apple' }, { value: 'pear' }, { value: 'orange' }, { value: 'grape' }, { value: 'banana' }];

<DropdownFiltered
  label="Fruits"
  items={items}
  name="fruit-sized"
  onChange={selection => selection && console.log(`You selected ${selection.value}`)}
  defaultSelectedItem={{ value: 'orange' }}
  inputProps={{
    disabled: true,
    placeholder: 'Select a fruit...',
    name: 'fruits6',
  }}
  size="short"
/>;
```
