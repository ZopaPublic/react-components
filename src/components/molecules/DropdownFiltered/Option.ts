import { colors, typography } from '../../../constants';
import styled, { css } from 'styled-components';
import { AppTheme } from '../../styles/Theme';

interface OptionProps extends React.HTMLAttributes<HTMLDivElement> {
  highLighted?: boolean;
  selected?: boolean;
}

const Option = styled.div<OptionProps & { theme: AppTheme }>`
  cursor: pointer;
  padding: 8px;
  line-height: 32px;
  font-family: ${({ theme }) => theme.typography.primary};
  font-size: ${typography.sizes.text.body};
  font-weight: 600;
  color: ${colors.greyDarkest};
  ${({ selected, highLighted }) =>
    (selected || highLighted) &&
    css`
      color: ${({ theme }) => theme.input.searchInput.options.hover.color};
      background-color: ${({ theme }) => theme.input.searchInput.options.hover.backgroundColor};
    `};
`;

export default Option;
