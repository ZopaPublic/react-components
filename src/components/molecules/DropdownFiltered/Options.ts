import styled from 'styled-components';
import { colors } from '../../../constants';
import { AppTheme } from '../../styles/Theme';

export interface OptionsListProps {
  optionsListMaxHeight?: string;
}

const Options = styled.div<OptionsListProps & { theme: AppTheme }>`
  z-index: 1;
  width: 100%;
  background: ${colors.white};
  border: ${({ theme }) => theme.input.searchInput.border};
  border-top: 0;
  position: absolute;
  overflow: auto;
  border-radius: ${({ theme }) => theme.input.searchInput.options.borderRadius};
  max-height: ${({ optionsListMaxHeight = 'initial' }) => optionsListMaxHeight};
  box-shadow: ${({ theme }) => theme.input.searchInput.boxShadow};
`;

export default Options;
