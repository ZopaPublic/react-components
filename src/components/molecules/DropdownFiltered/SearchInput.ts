import styled, { css } from 'styled-components';
import InputText from '../../atoms/InputText/InputText';
import { InputProps } from '../../types';
import { AppTheme } from '../../styles/Theme';

export interface SearchInputProps {
  isOpen?: boolean;
}

export const SearchInput = styled(InputText)<SearchInputProps & InputProps & { theme: AppTheme }>`
  input {
    ${({ hasError }) => !hasError && 'margin-bottom: 0'};
    ${({ isOpen }) =>
      isOpen &&
      css`
        box-shadow: ${({ theme }) => theme.input.searchInput.boxShadow};
        border-radius: ${({ theme }) => theme.input.searchInput.borderRadiusInput};
        border-color: ${({ theme }) => theme.input.searchInput.borderColor};
      `};
  }
  input + span {
    cursor: pointer;
  }
  input:disabled + span {
    cursor: not-allowed;
  }
`;
