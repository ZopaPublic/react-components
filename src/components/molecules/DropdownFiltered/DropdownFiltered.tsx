import Downshift, { ControllerStateAndHelpers, DownshiftProps } from 'downshift';
import React from 'react';
import styled from 'styled-components';
import { colors } from '../../../constants';
import ErrorMessage from '../../atoms/ErrorMessage/ErrorMessage';
import InputLabel from '../../atoms/InputLabel/InputLabel';
import SizedContainer from '../../layout/SizedContainer/SizedContainer';
import Icon from '../../atoms/Icon/Icon';
import { faSort } from '@fortawesome/free-solid-svg-icons';
import unbrandedChevron from '../../../content/images/unbranded-chevron.svg';
import { FieldProps, InputProps } from '../../types';
import { useThemeContext } from '../../styles/Theme';
import Option from './Option';
import Options, { OptionsListProps } from './Options';
import { SearchInputProps, SearchInput } from './SearchInput';

export interface DropdownItem {
  value: string;
}

export interface DropdownFilteredProps
  extends DownshiftProps<DropdownItem>,
    SearchInputProps,
    OptionsListProps,
    Omit<InputProps, 'onSelect' | 'onChange' | 'children'>,
    FieldProps {
  /**
   * Native props for the native label element.
   */
  labelProps?: React.LabelHTMLAttributes<HTMLLabelElement>;
  /**
   * Items for the options
   */
  items: DropdownItem[];
}

const FieldError = styled(ErrorMessage)`
  margin-top: 5px;
`;

export const SearchInputWrap = styled.div`
  position: relative;
`;

export const CustomIcon = styled.div<{ isOpen: boolean }>`
  background: transparent url(${unbrandedChevron}) no-repeat center;
  background-size: 60%;
  width: 100%;
  height: 100%;
  color: black;
  transform: ${({ isOpen }) => (isOpen ? 'rotate(180deg)' : 'none')};
`;

const DropdownFiltered = (props: DropdownFilteredProps) => {
  const {
    errorMessage,
    items = [],
    label,
    optionsListMaxHeight,
    name,
    labelProps,
    isValid,
    disabled,
    onChange,
    onFocus,
    inputSize,
    ...rest
  } = props;

  const searchMatch = (itemValue: string, inputValue: string) =>
    itemValue && itemValue.toLowerCase().includes(inputValue.toLowerCase());

  const theme = useThemeContext();

  return (
    <SizedContainer size={inputSize}>
      <Downshift itemToString={(item) => (item ? item.value : '')} {...props}>
        {({
          clearSelection,
          getInputProps,
          getItemProps,
          getLabelProps,
          getMenuProps,
          highlightedIndex,
          inputValue,
          isOpen,
          openMenu,
          closeMenu,
          selectedItem,
        }: ControllerStateAndHelpers<DropdownItem>) => {
          const filteredResults = items.filter(({ value }) => searchMatch(value, inputValue || ''));

          const showError = errorMessage && !isOpen;
          return (
            <div>
              {label && (
                <InputLabel {...getLabelProps({ ...labelProps })} htmlFor={`text-id-${name}`}>
                  {label}
                </InputLabel>
              )}
              <SearchInputWrap>
                <SearchInput
                  {...(getInputProps as any)({
                    // Types doesn't allow custom data-* attrs
                    onChange: (e: React.ChangeEvent<HTMLInputElement>) => {
                      if (e.target.value === '') {
                        clearSelection();
                      }
                    },
                    onFocus: () => {
                      openMenu();
                    },
                    ...rest,
                  })}
                  name={name}
                  id={`text-id-${name}`}
                  isValid={isValid}
                  hasError={showError}
                  isOpen={isOpen && !!filteredResults.length}
                  disabled={disabled}
                  theme={theme}
                  endIcon={
                    theme.input.searchInput.customIcon ? (
                      <CustomIcon
                        isOpen={isOpen}
                        onClick={() => {
                          if (!disabled) {
                            if (isOpen) {
                              closeMenu();
                            } else {
                              clearSelection(() => {
                                openMenu();
                              });
                            }
                          }
                        }}
                        aria-label={isOpen ? 'close.menu' : 'open.menu'}
                      />
                    ) : (
                      <Icon
                        variant={faSort}
                        rotation={isOpen ? 180 : undefined}
                        color={colors.grey}
                        onClick={() => {
                          if (!disabled) {
                            if (isOpen) {
                              closeMenu();
                            } else {
                              clearSelection(() => {
                                openMenu();
                              });
                            }
                          }
                        }}
                        aria-label={isOpen ? 'close.menu' : 'open.menu'}
                      />
                    )
                  }
                />
                {isOpen && !!filteredResults.length && (
                  <Options {...getMenuProps()} optionsListMaxHeight={optionsListMaxHeight} theme={theme}>
                    {filteredResults.map((item, index) => (
                      <Option
                        {...getItemProps({ index, item })}
                        key={item.value}
                        highLighted={highlightedIndex === index}
                        selected={selectedItem === item}
                        disabled={disabled}
                        theme={theme}
                      >
                        {item.value}
                      </Option>
                    ))}
                  </Options>
                )}
              </SearchInputWrap>
              {showError && <FieldError data-automation={`ZA.error-${name}`}>{errorMessage}</FieldError>}
            </div>
          );
        }}
      </Downshift>
    </SizedContainer>
  );
};

export default DropdownFiltered;
