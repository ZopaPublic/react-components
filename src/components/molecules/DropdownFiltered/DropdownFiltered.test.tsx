import React from 'react';
import { render } from '@testing-library/react';
import axe from '../../../../axe-helper';
import DropdownFiltered from './DropdownFiltered';

const items = [{ value: 'apple' }, { value: 'pear' }, { value: 'orange' }, { value: 'grape' }, { value: 'banana' }];

describe('<DropdownFiltered />', () => {
  it('renders the Dropdown with options with no a11y violations', async () => {
    const { container } = render(<DropdownFiltered items={items} label="Label" name="test" />);
    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });

  it('renders the DropdownFiltered with options and a default selected value', () => {
    const { container } = render(
      <DropdownFiltered initialSelectedItem={{ value: 'orange' }} items={items} name="test" />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
