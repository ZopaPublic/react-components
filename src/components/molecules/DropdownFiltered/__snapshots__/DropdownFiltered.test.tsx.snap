// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownFiltered /> renders the Dropdown with options with no a11y violations 1`] = `
.c1 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c0 {
  max-width: 100%;
  width: 100%;
}

.c3 {
  position: relative;
}

.c6 {
  position: absolute;
  top: 1px;
  bottom: 1px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 48px;
  color: #818F9B;
  right: 1px;
  border-top-right-radius: 8px;
  border-bottom-right-radius: 8px;
}

.c5 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  height: 50px;
  padding: 0 16px;
  padding-right: 60px;
  font-size: 16px;
  font-weight: 400;
  color: #4A545E;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
}

.c5:hover {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c5:focus {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E #009E8E;
}

.c5::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c5::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c5:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c5::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c5:disabled {
  -webkit-text-fill-color: #818F9B;
  opacity: 1;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  background-color: #F7F7F7;
  cursor: not-allowed;
}

.c4 input {
  margin-bottom: 0;
}

.c4 input + span {
  cursor: pointer;
}

.c4 input:disabled + span {
  cursor: not-allowed;
}

.c2 {
  position: relative;
}

<div
  class="c0"
>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    role="combobox"
  >
    <label
      class="c1"
      for="text-id-test"
      id="downshift-0-label"
    >
      Label
    </label>
    <div
      class="c2"
    >
      <div
        class="c3 c4"
      >
        <input
          aria-autocomplete="list"
          aria-labelledby="downshift-0-label"
          autocomplete="off"
          class="c5"
          id="text-id-test"
          name="test"
          value=""
        />
        <span
          class="c6"
        >
          <svg
            aria-hidden="true"
            aria-label="open.menu"
            class="svg-inline--fa fa-sort fa-w-10 "
            color="#818F9B"
            data-icon="sort"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<DropdownFiltered /> renders the DropdownFiltered with options and a default selected value 1`] = `
.c0 {
  max-width: 100%;
  width: 100%;
}

.c2 {
  position: relative;
}

.c5 {
  position: absolute;
  top: 1px;
  bottom: 1px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 48px;
  color: #818F9B;
  right: 1px;
  border-top-right-radius: 8px;
  border-bottom-right-radius: 8px;
}

.c4 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  height: 50px;
  padding: 0 16px;
  padding-right: 60px;
  font-size: 16px;
  font-weight: 400;
  color: #4A545E;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
}

.c4:hover {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c4:focus {
  border: 1px solid #009E8E;
  box-shadow: 0 0 4px 0 #009E8E #009E8E;
}

.c4::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c4::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c4:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c4::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c4:disabled {
  -webkit-text-fill-color: #818F9B;
  opacity: 1;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  background-color: #F7F7F7;
  cursor: not-allowed;
}

.c3 input {
  margin-bottom: 0;
}

.c3 input + span {
  cursor: pointer;
}

.c3 input:disabled + span {
  cursor: not-allowed;
}

.c1 {
  position: relative;
}

<div
  class="c0"
>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-1-label"
    role="combobox"
  >
    <div
      class="c1"
    >
      <div
        class="c2 c3"
      >
        <input
          aria-autocomplete="list"
          aria-labelledby="downshift-1-label"
          autocomplete="off"
          class="c4"
          id="text-id-test"
          name="test"
          value="orange"
        />
        <span
          class="c5"
        >
          <svg
            aria-hidden="true"
            aria-label="open.menu"
            class="svg-inline--fa fa-sort fa-w-10 "
            color="#818F9B"
            data-icon="sort"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
              fill="currentColor"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
